crates/compiler/specialize_types/src/mono_ir.rs

```diff
@@ -0,0 +1,372 @@
+        ret_type: MonoTypeId,
+    },
+
+    Lambda {
+        fn_type: MonoTypeId,
+        arguments: Slice<(MonoTypeId, MonoPatternId)>,
+        body: MonoExprId,
+        captured_symbols: Slice<(IdentId, MonoTypeId)>,
+        recursive: Recursive,
+    },
+
+    /// A record literal or a tuple literal.
+    /// These have already been sorted alphabetically.
+    Struct(NonEmptySlice<MonoExpr>),
+
+    /// Look up exactly one field on a record, tuple, or tag payload.
+    /// At this point we've already unified those concepts and have
+    /// converted (for example) record field names to indices, and have
+    /// also dropped all fields that have no runtime representation (e.g. empty records).
+    ///
+    /// In a later compilation phase, these indices will be re-sorted
+    /// by alignment and converted to byte offsets, but we in this
+    /// phase we aren't concerned with alignment or sizes, just indices.
+    StructAccess {
+        record_expr: MonoExprId,
+        record_type: MonoTypeId,
+        field_type: MonoTypeId,
+        field_id: MonoFieldId,
+    },
+
+    RecordUpdate {
+        record_type: MonoTypeId,
+        record_name: IdentId,
+        updates: Slice2<MonoFieldId, MonoExprId>,
+    },
+
+    /// Same as BigTag but with u8 discriminant instead of u16
+    SmallTag {
+        discriminant: u8,
+        tag_union_type: MonoTypeId,
+        args: Slice2<MonoTypeId, MonoExprId>,
+    },
+
+    /// Same as SmallTag but with u16 discriminant instead of u8
+    BigTag {
+        discriminant: u16,
+        tag_union_type: MonoTypeId,
+        args: Slice2<MonoTypeId, MonoExprId>,
+    },
+
+    Block {
+        stmts: Slice<MonoStmtId>,
+        final_expr: MonoExprId,
+    },
+
+    CompilerBug(Problem),
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct WhenBranch {
+    pub patterns: Slice<MonoPatternId>,
+    pub body: Slice<MonoStmtId>,
+    pub guard: Option<MonoExprId>,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum MonoPattern {
+    Identifier(IdentId),
+    As(MonoPatternId, IdentId),
+    StrLiteral(InternedStrId),
+    NumberLiteral(Number),
+    AppliedTag {
+        tag_union_type: MonoTypeId,
+        tag_name: IdentId,
+        args: Slice<MonoPatternId>,
+    },
+    StructDestructure {
+        struct_type: MonoTypeId,
+        destructs: Slice3<IdentId, MonoFieldId, DestructType>,
+    },
+    List {
+        elem_type: MonoTypeId,
+        patterns: Slice<MonoPatternId>,
+
+        /// Where a rest pattern splits patterns before and after it, if it does at all.
+        /// If present, patterns at index >= the rest index appear after the rest pattern.
+        /// For example:
+        ///   [ .., A, B ] -> patterns = [A, B], rest = 0
+        ///   [ A, .., B ] -> patterns = [A, B], rest = 1
+        ///   [ A, B, .. ] -> patterns = [A, B], rest = 2
+        /// Optionally, the rest pattern can be named - e.g. `[ A, B, ..others ]`
+        opt_rest: Option<(u16, Option<IdentId>)>,
```

Comment:
Why do you think we need to future proof? Personally, I prefer types to model the current reality. We can change it later and just follow the compiler.

Reply:
I was thinking of how easy it was to implement early returns (barring some minor bugs) in the `mono` code + codegen. Though I don't know if early returns were ever planned, the refcounting code Just Worked when I converted all early returns `can::Expr`s to `Stmt::Ret` naively.

With how complex Roc is at 300k LoC and growing, it's nice to make code work in a more generic way to avoid future contributors needing to figure out how the whole thing works before making changes. 

That said, it's minor, and I can see why we'd rather them need to understand how things work before making changes. I'm totally fine if we don't go for the "future-proof" approach.

Reply:
I think I'm in a similar boat. Sounds good to me to keep it as it currently is.